src/peft/tuners/boft/layer.py

```diff
@@ -0,0 +1,925 @@
+            n_random = torch.randint(0, N, (1,)).item()
+
+            # Create a mask with 1s for matrices to be replaced with identity and 0s otherwise
+            num_to_replace = int(self.p * D)
+            num_zeros = D - num_to_replace
+
+            # Generate a flat tensor with desired number of 1s and 0s
+            mask = torch.cat([torch.ones(num_to_replace, device=x.device), torch.zeros(num_zeros, device=x.device)])
+
+            # Shuffle and reshape the mask
+            mask = mask[torch.randperm(D)].view(1, D, 1, 1)
+
+            full_mask = torch.zeros(N, D, 1, 1, device=x.device)
+            full_mask[n_random] = mask
+
+            # Use the mask to combine original matrices and identity matrices
+            eye_matrix = torch.eye(H, device=x.device).repeat(N, D, 1, 1)
+            x = (1 - full_mask) * x + full_mask * eye_matrix
+        return x
+
+
+class BOFTLayer(BaseTunerLayer):
+    """
+    Implements the BOFT layer.
+    """
+    # All names of layers that may contain (trainable) adapter weights
+    adapter_layer_names = ("boft_R", "boft_s")
+    # All names of other parameters that may contain adapter-related parameters
+    other_param_names = ("boft_block_size", "boft_block_num", "boft_dropout")
+
+    def __init__(self, base_layer: nn.Module, **kwargs) -> None:
+        """
+        Initializes the BOFT layer.
+
+        Note, currently only support linear layer and convolutional layer, with further support for other layers to be added soon.
+
+        Parameters:
+        base_layer: the pretrained model layer
+        """
+        self.base_layer = base_layer
+        self.boft_block_size = {}
+        self.boft_block_num = {}
+        self.boft_dropout = nn.ModuleDict({})
+        self.boft_R = nn.ParameterDict({})
+        self.boft_s = nn.ParameterDict({})
+        # Mark the weight as unmerged
+        self._disable_adapters = False
+        self.merged_adapters = []
+        self.kwargs = kwargs
+
+        base_layer = self.get_base_layer()
+
+        if isinstance(base_layer, nn.Linear):
+            in_features, out_features = base_layer.in_features, base_layer.out_features
+        elif isinstance(base_layer, nn.Conv2d):
+            in_features, out_features = base_layer.in_channels, base_layer.out_channels
+        else:
+            raise ValueError(f"Unsupported layer type {type(base_layer)}")
+
+        self.in_features = in_features
+        self.out_features = out_features
+        
+        global fbd_cuda
+        fbd_cuda = load(
+            name="fbd_cuda",
+            sources=[f"{curr_dir}/fbd/fbd_cuda.cpp", f"{curr_dir}/fbd/fbd_cuda_kernel.cu"],
+            verbose=True,
+            # build_directory='/tmp/'  # for debugging
+            )
+        # extra_cuda_cflags = ['-std=c++14', '-ccbin=$$(which gcc-7)']) # cuda10.2 is not compatible with gcc9. Specify gcc 7
+        import fbd_cuda
+
+    def unscale_layer(self, scale=None) -> None:
```

Comment:
The stable diffusion unit tests are currently failing for me because they try to call `scale_layer`. If BOFT should work with diffusers, `scale_layer` should be added and `unscale_layer` get implemented properly.

Reply:
@BenjaminBossan We do want to use diffusers. What is the purpose of this function? Can we use two dummy functions?

Reply:
This method exists to apply the alpha scale parameter to the LoRA weights. This is important during the `forward` call when a diffusers user wants to scale LoRA with a weight != 1. It can also come into play during merging of the LoRA weights.

Here is an example of how it's called in diffusers:

https://github.com/huggingface/diffusers/blob/5f150c4cef2140342c66ef44ba8f36b5844a51b1/src/diffusers/models/unets/unet_2d_condition.py#L1191

Unscaling basically does the inverse, with the option to resetting the scale to the initial value of no argument is passed. See the two methods implemented for LoRA as reference:

https://github.com/huggingface/peft/blob/9119b780ebac7859db5753ebad50d94ba803c99c/src/peft/tuners/lora/layer.py#L237-L255

Reply:
Thanks for the clarification. I looked into the original paper of LoRA, BOFT should not support this kind of scaling, since it performs a multiplicative updated. Still, I have added the functions scale_layer, unscale_layer and set_scale functions to catch the function call and  leave a warning message. Fixed.